flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -0,0 +1,334 @@
+	}
+
+	@Override
+	public CompletableFuture<FlinkService> createInternalService(String clusterId) {
+		return createService(clusterId, this.internalServiceDecorators);
+	}
+
+	@Override
+	public CompletableFuture<FlinkService> createRestService(String clusterId) {
+		return createService(clusterId + Constants.FLINK_REST_SERVICE_SUFFIX, this.restServiceDecorators);
+	}
+
+	@Override
+	public void createFlinkMasterDeployment(ClusterSpecification clusterSpecification) {
+		FlinkDeployment deployment = new FlinkDeployment(this.flinkConfig);
+
+		for (Decorator<Deployment, FlinkDeployment> d : this.flinkMasterDeploymentDecorators) {
+			deployment = d.decorate(deployment);
+		}
+
+		deployment = new FlinkMasterDeploymentDecorator(clusterSpecification).decorate(deployment);
+		LOG.info("Create Flink Master deployment with spec: {}", deployment.getInternalResource().getSpec().toString());
+
+		this.internalClient.apps().deployments().inNamespace(this.nameSpace).create(
+			deployment.getInternalResource());
+	}
+
+	@Override
+	public void createTaskManagerPod(TaskManagerPodParameter parameter) {
+		FlinkPod pod = new FlinkPod(this.flinkConfig);
+
+		for (Decorator<Pod, FlinkPod> d : this.taskManagerPodDecorators) {
+			pod = d.decorate(pod);
+		}
+
+		pod = new TaskManagerPodDecorator(parameter).decorate(pod);
+		LOG.info("Create TaskManager pod with spec: {}", pod.getInternalResource().getSpec().toString());
+
+		this.internalClient.pods().inNamespace(this.nameSpace).create(pod.getInternalResource());
+	}
+
+	@Override
+	public void stopPod(String podName) {
+		this.internalClient.pods().withName(podName).delete();
+	}
+
+	@Override
+	@Nullable
+	public Endpoint getRestEndpoints(String clusterId) {
+		int restPort = this.flinkConfig.getInteger(RestOptions.PORT);
+		String serviceExposedType = flinkConfig.getString(KubernetesConfigOptions.REST_SERVICE_EXPOSED_TYPE);
+
+		// Return the service.namespace directly when use ClusterIP.
+		if (serviceExposedType.equals(KubernetesConfigOptions.ServiceExposedType.ClusterIP.toString())) {
+			return new Endpoint(clusterId + "." + nameSpace, restPort);
+		}
+
+		FlinkService restService = getRestService(clusterId);
+		if (restService == null) {
+			return null;
+		}
+		Service service = restService.getInternalResource();
+
+		String address = null;
+
+		if (service.getStatus() != null && (service.getStatus().getLoadBalancer() != null ||
+			service.getStatus().getLoadBalancer().getIngress() != null)) {
+			if (service.getStatus().getLoadBalancer().getIngress().size() > 0) {
+				address = service.getStatus().getLoadBalancer().getIngress().get(0).getIp();
+			} else {
+				address = this.internalClient.getMasterUrl().getHost();
+				restPort = getServiceNodePort(service, RestOptions.PORT);
+			}
+		} else if (service.getSpec().getExternalIPs() != null && service.getSpec().getExternalIPs().size() > 0) {
+			address = service.getSpec().getExternalIPs().get(0);
+		}
+		return new Endpoint(address, restPort);
```

Comment:
null `address ` (used to set `RestOptions.ADDRESS`) may cause troubles.
Maybe return null in this case for early error detection?

Reply:
Nice catch. I will add `return null` in the `else`.